File: typescript/Style/block.ts
// import * as _Config from "../type/config.js";
// import * as _File from "../type/file.js";
import * as _Style from "../type/style.js";
// import * as _Script from "../type/script.js";
// import * as _Cache from "../type/cache.js";
// import * as _Support from "../type/support.js";

import * as CACHE from "../data/cache.js";

import Use from "../utils/main.js";

const OPEN_CHARS = ["{", "[", "("];
const CLOSE_CHARS = ["}", "]", ")"];
const QUOTE_CHARS = ["`", "'", '"'];

export default function parseBlock(content: string, blockArrays = false) {
	content += ";";
	const length = content.length;

	let keyStart = 0,
		valStart = 0,
		deviance = 0,
		quote = "",
		key = "",
		isProp = true;

	const result: _Style.ParsedResult = {
		assign: [],
		attachment: [],
		constants: {},
		XatProps: [],
		atProps: {},
		Xproperties: [],
		properties: {},
		XatRules: [],
		atRules: {},
		Xnested: [],
		nested: {},
		Xclasses: [],
		classes: {},
		Xflats: [],
		flats: {},
		XallBlocks: [],
		allBlocks: {},
	};

	for (let index = 0; index < length; index++) {
		const ch = content[index];
		if (ch === "\\") {
			index++;
			continue;
		}
		if (QUOTE_CHARS.includes(ch)) {
			if (quote === "") {
				quote = ch;
			} else if (quote === ch) {
				quote = "";
			}
		}

		if (quote === "") {
			if (CLOSE_CHARS.includes(ch)) { deviance--; }

			if (deviance === 0) {
				switch (ch) {
					case "{":
						isProp = false;
						key = Use.string.minify(content.slice(keyStart, index));
						valStart = index + 1;
						break;
					case ":":
						key = Use.string.minify(content.slice(keyStart, index));
						valStart = index + 1;
						break;
					case "}":
					case ";": {
						const value = Use.string.minify(content.slice(valStart, index));
						if (isProp) {
							if (key.length > 0) {
								if (key.startsWith("--")) { result.constants[key] = value; }
								result.properties[key] = value;
								if (blockArrays) { result.Xproperties.push([key, value]); }
							} else if (value[0] === "@") {
								const firstSpaceIndex = value.indexOf(" ");
								const spaceIndex =
									firstSpaceIndex < 0 ? value.length : firstSpaceIndex;
								const directive = value.slice(0, spaceIndex);

								switch (directive) {
									case CACHE.ROOT.customAtrules["attach"]:
										result.attachment.push(
											...Use.string.zeroBreaks(value.slice(spaceIndex)),
										);
										break;
									case CACHE.ROOT.customAtrules["assign"]:
										result.assign.push(
											...Use.string.zeroBreaks(value.slice(spaceIndex)),
										);
										break;
									default:
										result.atProps[value] = "";
										if (blockArrays) { result.XatProps.push([value, ""]); }
								}
							} else {
								const breaks = Use.string.zeroBreaks(value);
								switch (breaks[0]) {
									case CACHE.ROOT.customOperations["attach"]:
										breaks.shift();
										result.attachment.push(...breaks);
										break;
									case CACHE.ROOT.customOperations["assign"]:
										breaks.shift();
										result.assign.push(...breaks);
										break;
								}
							}
						} else {
							switch (key[0]) {
								case "@":
									result.atRules[key] = value;
									if (blockArrays) { result.XatRules.push([key, value]); }
									break;
								case "&":
									result.nested[key] = value;
									if (blockArrays) { result.Xnested.push([key, value]); }
									break;
								case ".":
									result.classes[key] = value;
									if (blockArrays) { result.Xclasses.push([key, value]); }
									break;
								default:
									result.flats[key] = value;
									if (blockArrays) { result.Xflats.push([key, value]); }
							}
							result.allBlocks[key] = value;
							if (blockArrays) { result.XallBlocks.push([key, value]); }
						}
						keyStart = index + 1;
						valStart = index + 1;
						key = "";
						isProp = true;
					}
				}
			}

			if (OPEN_CHARS.includes(ch)) { deviance++; }
		}
	}

	return result;
}

File: go/style/block.go
package style

import (
	_cache_ "main/cache"
	_Cursor_ "main/class/Cursor"
	_utils_ "main/utils"
	_slices_ "slices"
	_strings_ "strings"
)

type block_Parse_retype struct {
	Assign      []string
	Attachment  []string
	Variables   map[string]string
	XatProps    [][2]string
	AtProps     map[string]string
	Xproperties [][2]string
	Properties  map[string]string
	XatRules    [][2]string
	AtRules     map[string]string
	Xnested     [][2]string
	Nested      map[string]string
	Xclasses    [][2]string
	Classes     map[string]string
	Xflats      [][2]string
	Flats       map[string]string
	XallBlocks  [][2]string
	AllBlocks   map[string]string
}

func Block_Parse(content string, blockInArrays bool) block_Parse_retype {
	content += ";"

	result := block_Parse_retype{
		Assign:      []string{},
		Attachment:  []string{},
		Variables:   map[string]string{},
		XatProps:    [][2]string{},
		AtProps:     map[string]string{},
		Xproperties: [][2]string{},
		Properties:  map[string]string{},
		XatRules:    [][2]string{},
		AtRules:     map[string]string{},
		Xnested:     [][2]string{},
		Nested:      map[string]string{},
		Xclasses:    [][2]string{},
		Classes:     map[string]string{},
		Xflats:      [][2]string{},
		Flats:       map[string]string{},
		XallBlocks:  [][2]string{},
		AllBlocks:   map[string]string{},
	}

	key := ""
	var awaitBrace rune = 0
	braceTrack := []rune{}
	keyStart := 0
	valStart := 0
	deviance := 0
	isProp := true
	cursor := _Cursor_.Construct(content + ";")

	for ch, streaming := cursor.Active.Char, cursor.Streaming; streaming; ch, streaming = cursor.Increment() {

		if ch == '\\' {
			cursor.Increment()
			continue
		} else if awaitBrace == ch {
			deviance = len(braceTrack) - 1
			braceTrack = braceTrack[:deviance]
			if deviance > 0 {
				awaitBrace = braceTrack[deviance-1]
			} else {
				awaitBrace = 0
			}
		} else if !_slices_.Contains(_utils_.Refer.WatchQuotes, awaitBrace) &&
			_slices_.Contains(_utils_.Refer.OpenBraces, ch) || _slices_.Contains(_utils_.Refer.WatchQuotes, ch) {
			awaitBrace = _utils_.Refer.BracePair[ch]
			braceTrack = append(braceTrack, awaitBrace)
			deviance = len(braceTrack)
		}

		if deviance == 1 && cursor.Active.Char == '{' {
			isProp = false
			key = _utils_.String_Minify(content[keyStart : cursor.Active.Marker-1])
			valStart = cursor.Active.Marker + 1
		} else if deviance != 0 {
			continue
		} else {
			switch cursor.Active.Char {
			case ':':
				key = _utils_.String_Minify(content[keyStart : cursor.Active.Marker-1])
				valStart = cursor.Active.Marker + 1
			case '}':
				fallthrough
			case ';':
				{
					value := _utils_.String_Minify(content[valStart : cursor.Active.Marker-1])
					if isProp {
						if len(key) > 0 {
							if _strings_.HasPrefix(key, "--") {
								result.Variables[key] = value
							}
							result.Properties[key] = value
							if blockInArrays {
								result.Xproperties = append(result.Xproperties, [2]string{key, value})
							}
						} else if len(value) > 0 && value[0] == '@' {
							spaceIndex := _strings_.Index(value, " ")
							if spaceIndex < 0 {
								spaceIndex = len(value)
							}
							directive := value[0:spaceIndex]

							switch directive {
							case _cache_.Root.CustomAtrules["attach"]:
								breaks := _utils_.String_ZeroBreaks(value[spaceIndex:], []rune{' ', '\n', ','})
								result.Attachment = append(result.Attachment, breaks...)
							case _cache_.Root.CustomAtrules["assign"]:
								breaks := _utils_.String_ZeroBreaks(value[spaceIndex:], []rune{' ', '\n', ','})
								result.Assign = append(result.Assign, breaks...)
							default:
								result.AtProps[value] = ""
								if blockInArrays {
									result.XatProps = append(result.XatProps, [2]string{value, ""})
								}
							}
						} else {
							breaks := _utils_.String_ZeroBreaks(value, []rune{' ', '\n', ','})
							if len(breaks) > 0 {
								if breaks[0] == string(_cache_.Root.CustomOperations["attach"]) {
									result.Attachment = append(result.Attachment, breaks[1:]...)
								} else if (breaks[0]) == string(_cache_.Root.CustomOperations["assign"]) {
									result.Assign = append(result.Assign, breaks[1:]...)
								}
							}
						}
					} else if len(key) > 0 {
						switch rune(key[0]) {
						case '@':
							result.AtRules[key] = value
							if blockInArrays {
								result.XatRules = append(result.XatRules, [2]string{key, value})
							}
						case '&':
							result.Nested[key] = value
							if blockInArrays {
								result.Xnested = append(result.Xnested, [2]string{key, value})
							}
						case '.':
							result.Classes[key] = value
							if blockInArrays {
								result.Xclasses = append(result.Xclasses, [2]string{key, value})
							}
						default:
							result.Flats[key] = value
							if blockInArrays {
								result.Xflats = append(result.Xflats, [2]string{key, value})
							}
						}
						result.AllBlocks[key] = value
						if blockInArrays {
							result.XallBlocks = append(result.XallBlocks, [2]string{key, value})
						}
					}
					keyStart = cursor.Active.Marker
					valStart = cursor.Active.Marker
					key = ""
					isProp = true
				}
			}
		}
	}

	return result
}
